Doc/howto/multiple-interpreters.rst

```diff
@@ -0,0 +1,1259 @@
+
+In the same way, running code in an interpreter does not reset that
+interpreter.  The next time you run code in that interpreter, the
+:mod:`!__main__` module will be in exactly the state in which you
+left it::
+
+    from concurrent import interpreters
+
+    interp = interpreters.create()
+
+    interp.exec("""if True:
+        answer = 42
+        """)
+    interp.exec("""if True:
+        assert answer == 42
+        """)
+
+Similarly::
+
+    from concurrent import interpreters
+
+    def set(value):
+        assert __name__ == '<fake __main__>', __name__
+        global answer
+        answer = value
+
+    def check(expected):
+        assert answer == expected
+
+    if __name__ == '__main__':
+        interp = interpreters.create()
+        interp.call(set, 100)
+        interp.call(check, 100)
+
+
+You can take advantage of this to prepare an interpreter ahead of time
+or use it incrementally::
+
+    from textwrap import dedent
+    from concurrent import interpreters
+
+    interp = interpreters.create()
+
+    # Prepare the interpreter.
+    interp.exec(dedent("""
+        # We will need this later.
+        import math
+
+        # Initialize the value.
+        value = 1
+
+        def double(val):
+            return val + val
+        """))
+
+    # Do the work.
+    for _ in range(9):
+        interp.exec(dedent("""
+            assert math.factorial(value + 1) >= double(value)
+            value = double(value)
+            """))
+
+    # Show the result.
+    interp.exec('print(value)')
+    # prints: 1024
+
+In case you're curious, in a little while we'll look at how to pass
+data in and out of an interpreter (instead of just printing things).
+
+Sometimes you might also have values that you want to use in a later
+script in another interpreter.  We'll look more closely at that
+case in a little while: :ref:`interp-script-args`.
+
+Running in a Thread
+-------------------
+
+A key point is that, in a new Python process, the runtime doesn't
+create a new thread just to run the target script.  It uses the main
+thread the process started with.
+
+In the same way, running code in another interpreter doesn't create
+(or need) a new thread.  It uses the current OS thread, switching
+to the interpreter long enough to run the code.
+
+To actually run in a new thread, you can do it explicitly::
+
+    from concurrent import interpreters
+    import threading
+
+    interp = interpreters.create()
+
+    def run():
+        interp.exec("print('spam!')")
+    t = threading.Thread(target=run)
+    t.start()
+    t.join()
+
+    def run():
+        def script():
+            print('spam!')
+        interp.call(script)
+    t = threading.Thread(target=run)
+    t.start()
+    t.join()
+
+There's also a helper method for that::
+
+    from concurrent import interpreters
+
+    def script():
+        print('spam!')
+
+    if __name__ == '__main__':
+        interp = interpreters.create()
+        t = interp.call_in_thread(script)
```

Comment:
Off-topic, but would it be possible to add the `daemon` parameter to the helper method?

Reply:
I want to keep the helper super simple.  If someone wants a daemon thread then they can do it manually without much effort.